File: 03-coding-patterns/examples/08-longest-substring-no-repeat/solution.ts
// Longest Substring Without Repeating Characters — TypeScript
function lengthOfLongestSubstring(s: string): number {
    const seen = new Map<string, number>();
    let l = 0, best = 0;
    for (let r = 0; r < s.length; r++) {
        const ch = s[r]!;
        const prev = seen.get(ch);
        if (prev !== undefined && prev >= l) {
            l = prev + 1;
        }
        seen.set(ch, r);
        best = Math.max(best, r - l + 1);
    }
    return best;
}

function check(ok: boolean, label: string): void {
    if (!ok) { console.error(`FAIL: ${label}`); process.exit(1); }
}

const cases: Array<[string, number]> = [
    ["abcabcbb", 3],
    ["bbbbb", 1],
    ["pwwkew", 3],
    ["", 0],
    ["a", 1],
    ["dvdf", 3],
    ["abba", 2],
];
for (const [s, want] of cases) {
    check(lengthOfLongestSubstring(s) === want, `s="${s}"`);
}
console.log("PASS");

File: 03-coding-patterns/examples/08-longest-substring-no-repeat/solution.py
"""Longest Substring Without Repeating Characters — Python 3.12."""


def length_of_longest_substring(s: str) -> int:
    seen: dict[str, int] = {}
    l = 0
    best = 0
    for r, ch in enumerate(s):
        if (prev := seen.get(ch, -1)) >= l:
            l = prev + 1
        seen[ch] = r
        best = max(best, r - l + 1)
    return best


if __name__ == "__main__":
    cases = [
        ("abcabcbb", 3),
        ("bbbbb", 1),
        ("pwwkew", 3),
        ("", 0),
        ("a", 1),
        ("dvdf", 3),
        ("abba", 2),
    ]
    for s, want in cases:
        got = length_of_longest_substring(s)
        assert got == want, f"got {got}, want {want} for {s!r}"
    print("PASS")
